transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -93,132 +108,269 @@ public Channel channel() {
+        final AbstractChannelHandlerContext newCtx;
+        final EventExecutor executor;
+        final boolean inEventLoop;
+        synchronized (this) {
+            checkMultiplicity(handler);
+
+            if (group != null) {
+                invoker = findInvoker(group);
+            }
+
+            newCtx = new DefaultChannelHandlerContext(this, invoker, filterName(name, handler), handler);
+            executor = executorSafe(invoker);
+            if (executor == null) {
+                addFirst0(newCtx);
+                callHandlerCallbackLater(newCtx, true);
+                return this;
+            }
+            inEventLoop = executor.inEventLoop();
+            if (inEventLoop) {
+                addFirst0(newCtx);
+            }
+        }
+
+        if (inEventLoop) {
+            callHandlerAdded0(newCtx);
+        } else {
+            waitForFuture(executor.submit(new OneTimeTask() {
+                @Override
+                public void run() {
+                    synchronized (DefaultChannelPipeline.this) {
+                        addFirst0(newCtx);
+                    }
+                    callHandlerAdded0(newCtx);
+                }
+            }));
+        }
         return this;
     }
 
     private void addFirst0(AbstractChannelHandlerContext newCtx) {
-        checkMultiplicity(newCtx);
-
         AbstractChannelHandlerContext nextCtx = head.next;
         newCtx.prev = head;
         newCtx.next = nextCtx;
         head.next = newCtx;
         nextCtx.prev = newCtx;
-
-        callHandlerAdded(newCtx);
     }
 
     @Override
     public ChannelPipeline addLast(String name, ChannelHandler handler) {
-        return addLast((ChannelHandlerInvoker) null, name, handler);
+        return addLast(null, null, name, handler);
     }
 
     @Override
-    public synchronized ChannelPipeline addLast(EventExecutorGroup group, String name, ChannelHandler handler) {
-        name = filterName(name, handler);
-        addLast0(new DefaultChannelHandlerContext(this, findInvoker(group), name, handler));
-        return this;
+    public ChannelPipeline addLast(EventExecutorGroup group, String name, ChannelHandler handler) {
+        return addLast(group, null, name, handler);
     }
 
     @Override
-    public synchronized ChannelPipeline addLast(ChannelHandlerInvoker invoker, String name, ChannelHandler handler) {
-        name = filterName(name, handler);
-        addLast0(new DefaultChannelHandlerContext(this, invoker, name, handler));
+    public ChannelPipeline addLast(ChannelHandlerInvoker invoker, String name, ChannelHandler handler) {
+        return addLast(null, invoker, name, handler);
+    }
+
+    private ChannelPipeline addLast(EventExecutorGroup group, ChannelHandlerInvoker invoker,
+                                    String name, ChannelHandler handler) {
+        assert !(group != null && invoker != null);
```

Comment:
- Could we simplify this: `assert group == null || invoker == null : "either group or invoker must be null"`
- I see the same assertion in multiple places. Could we extract this to a dedicated method?